wpilibc/src/main/native/cpp/LEDPattern.cpp

```diff
@@ -18,32 +19,54 @@
 
 using namespace frc;
 
-LEDPattern::LEDPattern(LEDPatternFn impl) : m_impl(std::move(impl)) {}
+LEDPattern::LEDPattern(std::function<void(frc::LEDPattern::LEDReader,
+                                          std::function<void(int, frc::Color)>)>
+                           impl)
+    : m_impl(std::move(impl)) {}
+
+void LEDPattern::ApplyTo(LEDPattern::LEDReader reader,
+                         std::function<void(int, frc::Color)> writer) const {
+  m_impl(reader, writer);
+}
 
 void LEDPattern::ApplyTo(std::span<AddressableLED::LEDData> data,
-                         LEDWriterFn writer) const {
-  m_impl(data, writer);
+                         std::function<void(int, frc::Color)> writer) const {
+  ApplyTo(LEDPattern::LEDReader{[=](size_t i) { return data[i]; }, data.size()},
+          writer);
 }
 
 void LEDPattern::ApplyTo(std::span<AddressableLED::LEDData> data) const {
   ApplyTo(data, [&](int index, Color color) { data[index].SetLED(color); });
 }
 
+// void LEDPattern:ApplyTo(std::range::views::reversed<AddressableLED::LEDData>
+// data) const {
+//   ApplyTo(data, [&](int index, Color color) { data[index].SetLED(color); });
+// }
+
 LEDPattern LEDPattern::Reversed() {
   return LEDPattern{[self = *this](auto data, auto writer) {
-    self.ApplyTo(data, [&](int i, Color color) {
-      writer((data.size() - 1) - i, color);
-    });
+    self.ApplyTo(
+        LEDPattern::LEDReader{[=](auto i) { return data[data.size() - 1 - i]; },
+                              data.size()},
+        [&](int i, Color color) { writer((data.size() - 1) - i, color); });
   }};
 }
 
 LEDPattern LEDPattern::OffsetBy(int offset) {
-  return LEDPattern{[=, self = *this](auto data, auto writer) {
-    self.ApplyTo(data, [&data, &writer, offset](int i, Color color) {
-      int shiftedIndex =
-          frc::FloorMod(i + offset, static_cast<int>(data.size()));
-      writer(shiftedIndex, color);
-    });
+  return LEDPattern{[=, self = *this](LEDPattern::LEDReader data, auto writer) {
+    self.ApplyTo(LEDPattern::LEDReader{[=, d = std::move(data)](auto i) {
```

Comment:
Is a std::move is necessary, or would a value capture in the lambda would be enough?

Reply:
Based on my reading of https://en.cppreference.com/w/cpp/language/lambda, using a value capture would copy `LEDReader`, which would copy the `std::function` member, whereas using the `std::move()` moves the (by-value) parameter, which I think is slightly preferable. I'll let others weigh in on whether this is correct and if the difference is significant.

Reply:
As I understand it, std::move could make the original owner invalid, which would be fine if the original owners are short lived intermediate objects. But I don't think that's actually the case here (or at least isn't guaranteed). It's entirely possible to do `LEDPattern::Gradient(...).Reversed()`, where the gradient is really a member of the reversed pattern, or `auto gradient = LEDPattern::Gradient(...); auto reversedGradient = gradient.Reversed()`, where the original gradient is still referenced and presumably used

Reply:
Yes, `std::move()` can make the original owner invalid. However, as I understand it, since `LEDPattern::LEDReader` data is pass by copy, the "original owner" is our copy of the argument, not what was passed in.

That said, I just noticed that we do reference `data` elsewhere in the outer lambda bodies, so I think it does need to be a by-copy capture.

> Is a std::move is necessary, or would a value capture in the lambda would be enough?

After re-reading this, I have a question- What concerns did you have about a value capture?